librad2/src/git/mod.rs

```diff
@@ -0,0 +1,353 @@
+        let mut contrib = meta::Contributor::new();
+        contrib.profile = Some(meta::ProfileRef::UserProfile(profile.clone()));
+        let cid = commit_contributor_meta(
+            sources,
+            &head,
+            &mut pgp_key,
+            "Radicle: initial contributor metadata",
+            contrib,
+        )?;
+        let mut contrib_branch = sources.branch(
+            CONTRIBUTOR_METADATA_BRANCH,
+            &sources.find_commit(cid)?,
+            true,
+        )?;
+
+        // Create a remote in our state dir
+        let res = register_project(paths, &pid, sources);
+
+        // Clean up local stuff
+        let _ = proj_branch.delete();
+        let _ = contrib_branch.delete();
+
+        res.map(|_| pid)
+    }
+
+    pub fn metadata(&self) -> Result<meta::Project, Error> {
+        let blob = {
+            self.0
+                .find_branch(PROJECT_METADATA_BRANCH, git2::BranchType::Local)?
+                .get()
+                .peel_to_tree()?
+                .get_name(PROJECT_METADATA_FILE)
+                .expect("Missing project.json")
+                .to_object(&self.0)?
+                .peel_to_blob()
+        }?;
+        let meta = serde_json::from_slice(blob.content())?;
+        Ok(meta)
+    }
+}
+
+fn commit_project_meta(
+    repo: &git2::Repository,
+    parent: &git2::Commit,
+    pgp_key: &mut pgp::Key,
+    msg: &str,
+    meta: meta::Project,
+) -> Result<git2::Oid, Error> {
+    commit_meta(repo, parent, pgp_key, msg, meta, PROJECT_METADATA_FILE)
+}
+
+fn commit_contributor_meta(
+    repo: &git2::Repository,
+    parent: &git2::Commit,
+    pgp_key: &mut pgp::Key,
+    msg: &str,
+    meta: meta::Contributor,
+) -> Result<git2::Oid, Error> {
+    commit_meta(repo, parent, pgp_key, msg, meta, CONTRIBUTOR_METADATA_FILE)
+}
+
+fn commit_meta<M>(
+    repo: &git2::Repository,
+    parent: &git2::Commit,
+    pgp_key: &mut pgp::Key,
+    msg: &str,
+    meta: M,
+    filename: &str,
+) -> Result<git2::Oid, Error>
+where
+    M: Serialize,
+{
+    let blob_oid = {
+        let mut blob = repo.blob_writer(None)?;
+        let mut ser = serde_json::Serializer::with_formatter(&mut blob, CanonicalFormatter::new());
+        meta.serialize(&mut ser)?;
+        blob.commit()?
+    };
+
+    let tree = {
+        let mut builder = repo.treebuilder(None)?;
+        builder.insert(filename, blob_oid, 0o100_644)?;
+        let oid = builder.write()?;
+        repo.find_tree(oid)?
+    };
+
+    let author = {
+        let addr = pgp_key
+            .userids()
+            .nth(0)
+            .ok_or(Error::InvalidKey)
+            .and_then(|binding| {
+                binding
+                    .userid()
+                    .address()
+                    .map_err(|_| Error::InvalidKey)
+                    .and_then(|addr| addr.ok_or(Error::InvalidKey))
+            })?;
+
+        // TODO: Check what happens if not set - is this returned as `Error`?
```

Comment:
I believe it errors. I came across this error during a CI test when I was doing git stuff in Rust code.